[Gradient Compression] Export sizes, along with length and offset of each variable to GradBucket for PowerSGD

wayi1 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#47203 [Gradient Compression] Export sizes, along with length and offset of each variable to GradBucket for PowerSGD**

1. Create a new field in BucketReplica to store sizes info for each variable.
2. Export sizes list, along with lengths and offsets to GradBuceket.

These fields are needed for PowerSGD.

Original PR issue: Investigate Applying PowerSGD to Communication Hook for Gradient Compression #47202

Differential Revision: [D24644137](https://our.internmc.facebook.com/intern/diff/D24644137/)

Commit:
[Gradient Compression] Export sizes, along with length and offset of each variable to GradBucket for PowerSGD

1. Create a new field in BucketReplica to store sizes info for each variable.
2. Export sizes list, along with lengths and offsets to GradBuceket.

These fields are needed for PowerSGD.

Original PR issue: Investigate Applying PowerSGD to Communication Hook for Gradient Compression #47202

Differential Revision: [D24644137](https://our.internmc.facebook.com/intern/diff/D24644137/)

[ghstack-poisoned]

rohan-varma commented on torch/csrc/distributed/c10d/comm.h:
```diff
@@ -33,14 +41,30 @@ class GradBucket {
     return tensors_;
   }
 
+  const std::vector<size_t>& getOffsets() const {
```
Nit: Add a comment describing each of these public APIs.

rohan-varma commented on torch/csrc/distributed/c10d/comm.h:
```diff
@@ -33,14 +41,30 @@ class GradBucket {
   }
 
+  const std::vector<size_t>& getOffsets() const {
+    return offsets_;
```
what are offsets in this context?

SciPioneer replied:
Start index of the variable in the flattened 1D tensor. Added a comment to describe this API.

rohan-varma commented on torch/csrc/distributed/c10d/reducer.cpp:
```diff
@@ -712,7 +712,13 @@ void Reducer::mark_bucket_ready(size_t bucket_index) {
-      GradBucket grad_bucket(tensors);
+      GradBucket grad_bucket(
+          tensors,
+          // Since currently we do not support single-process multiple-device
```
Do we need to `TORCH_CHECK` this assumption here or do we do that elsewhere already?

SciPioneer replied:
`TORCH_CHECK` is already done at the comm hook registration time, i.e., register_comm_hook method. This should be sufficient.

rohan-varma commented on torch/csrc/distributed/c10d/reducer.cpp:
```diff
@@ -785,6 +792,14 @@ void Reducer::initialize_buckets(
         at::TensorOptions options;
         size_t offset = 0;
 
+        // Reserve enough space for the per-variable fields stored in bucket
+        // replica for efficiency.
+        num_variables += bucket_indices[bucket_index].size();
+        replica.variables.reserve(num_variables);
+        replica.offsets.reserve(num_variables);
+        replica.lengths.reserve(num_variables);
+        replica.sizes_vec.reserve(num_variables);
+
```
Not completely sure I follow this logic. Don't we just want to reserve `bucket_indices[bucket_index].size()` for each replica? Why are we incrementing `num_variables` and using that as the argument into `reserve`?

SciPioneer replied:
You are right. I shouldn't increment `num_variables ` here. This wrong logic was not captured before because there is only one replica in this mode. Thanks!

rohan-varma reviewed (approved):
LGTM, but please address a few of the comments inline. Thanks!

rohan-varma commented on torch/csrc/distributed/c10d/comm.h:
```diff
@@ -33,14 +41,30 @@ class GradBucket {
+    return lengths_;
+  }
+
+  const std::vector<c10::IntArrayRef>& getSizesVec() const {
```
What's the difference here between lengths and sizes, and why is one `size_t` type and the other is `IntArrayRef` type?

SciPioneer replied:
`length` is the total 1D length like after flattening.
`sizes` is multi-dimensional shape info. Variable.sizes is similar to Tensor.shape.

Added a comment to describe the API.

Commit:
Update on "[Gradient Compression] Export sizes, along with length and offset of each variable to GradBucket for PowerSGD"


1. Create a new field in BucketReplica to store sizes info for each variable.
2. Export sizes list, along with lengths and offsets to GradBuceket.

These fields are needed for PowerSGD.

Original PR issue: Investigate Applying PowerSGD to Communication Hook for Gradient Compression #47202

Differential Revision: [D24644137](https://our.internmc.facebook.com/intern/diff/D24644137/)

[ghstack-poisoned]

Commit:
Update on "[Gradient Compression] Export sizes, along with length and offset of each variable to GradBucket for PowerSGD"


1. Create a new field in BucketReplica to store sizes info for each variable.
2. Export sizes list, along with lengths and offsets to GradBuceket.

These fields are needed for PowerSGD.

Original PR issue: Investigate Applying PowerSGD to Communication Hook for Gradient Compression #47202

Differential Revision: [D24644137](https://our.internmc.facebook.com/intern/diff/D24644137/)

[ghstack-poisoned]

pritamdamania87 commented on torch/csrc/distributed/c10d/init.cpp:
```diff
@@ -179,7 +188,19 @@ PyObject* c10d_init(PyObject* _unused, PyObject* noargs) {
-           )");
+           )")
+      .def(
+          "get_offsets",
+          &::c10d::GradBucket::getOffsets,
+          py::call_guard<py::gil_scoped_release>())
+      .def(
+          "get_lengths",
+          &::c10d::GradBucket::getLengths,
+          py::call_guard<py::gil_scoped_release>())
+      .def(
+          "get_sizes_list",
+          &::c10d::GradBucket::getSizesVec,
+          py::call_guard<py::gil_scoped_release>());
```
Would be nice to add docs for all of these similar to `get_tensors`
